oak-blob-plugins/src/main/java/org/apache/jackrabbit/oak/plugins/blob/datastore/AbstractDataStoreService.java

```diff
@@ -60,70 +67,103 @@
+    private static final Logger log = LoggerFactory.getLogger(AbstractDataStoreService.class);
 
     @Reference
     private StatisticsProvider statisticsProvider;
 
-    private DataStoreBlobStore dataStore;
+    protected Closer closer = Closer.create();
 
     protected void activate(ComponentContext context, Map<String, Object> config) throws RepositoryException {
         // change to mutable map. may be modified in createDS call
         config = Maps.newHashMap(config);
+
         DataStore ds = createDataStore(context, config);
+        registerDataStore(context, config, ds, getStatisticsProvider(), getDescription(), closer);
+    }
+
+    protected void deactivate() throws DataStoreException {
+        if (null != closer) {
+            synchronized (this) {
+                if (null != closer) {
+                    closeQuietly(closer);
+                    closer = null;
+                }
+            }
+        }
+    }
+
+    protected abstract DataStore createDataStore(ComponentContext context, Map<String, Object> config);
+
+    protected StatisticsProvider getStatisticsProvider(){
+        return statisticsProvider;
+    }
+
+    public static DataStore registerDataStore(
+            @Nonnull ComponentContext context,
+            @Nonnull Map<String, Object> config,
+            @Nullable DataStore ds,
+            @Nonnull StatisticsProvider statisticsProvider,
+            @Nonnull String[] description,
+            @Nonnull Closer closer
+    ) throws RepositoryException {
+        if (null == ds) {
+            // Deferred registration - child class should call registerDataStore later
+            return ds;
+        }
+
+        Closeables closeables = new Closeables(closer);
+
         boolean encodeLengthInId = PropertiesUtil.toBoolean(config.get(PROP_ENCODE_LENGTH), true);
         int cacheSizeInMB = PropertiesUtil.toInteger(config.get(PROP_CACHE_SIZE), DataStoreBlobStore.DEFAULT_CACHE_SIZE);
 
         String homeDir = lookup(context, PROP_HOME);
         if (config.containsKey(PATH) && !Strings.isNullOrEmpty((String) config.get(PATH))) {
             log.info("Initializing the DataStore with path [{}]", config.get(PATH));
+            homeDir = (String) config.get(PATH);
         }
         else if (homeDir != null) {
             log.info("Initializing the DataStore with homeDir [{}]", homeDir);
         }
         PropertiesUtil.populate(ds, config, false);
         ds.init(homeDir);
 
-        BlobStoreStats stats = new BlobStoreStats(getStatisticsProvider());
-        this.dataStore = new DataStoreBlobStore(ds, encodeLengthInId, cacheSizeInMB);
-        this.dataStore.setBlobStatsCollector(stats);
+        DataStoreBlobStore dataStore = new DataStoreBlobStore(ds, encodeLengthInId, cacheSizeInMB);
+        closeables.add(dataStore);
+        BlobStoreStats stats = new BlobStoreStats(statisticsProvider);
+        dataStore.setBlobStatsCollector(stats);
+
```

Comment:
TODO - We don't have the ability now to track data store stats for each of the delegate which should be added in some form. Either the BlobStoreStats should be enhanced to map stats for each delegate or provide each DataStore the ability to do that.